{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# РОЗДІЛ 6. ВИНЯТКИ\n",
    "\n",
    "## 6.1. Загальні поняття\n",
    "\n",
    "**Винятки** - це сповіщення інтерпретатора, порушувані в разі виникнення помилки в програмному коді або при настанні якої- небудь події. Якщо в коді не передбачено оброблення винятків, то програма переривається і виводиться повідомлення про помилку.\n",
    "\n",
    "Нагадаємо, що існує три типи помилок в програмі:\n",
    "\n",
    "**Синтаксичні** - це помилки в імені оператора або функції, невідповідність закриваючих та відкриваючих лапок і т.д. Тобто помилки в синтаксисі мови. Як правило, інтерпретатор попередить про наявність помилки, а програма не виконуватиметься зовсім. Приклад синтаксичної помилки:    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "EOL while scanning string literal (<ipython-input-1-8b05592a5f1a>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-1-8b05592a5f1a>\"\u001b[1;36m, line \u001b[1;32m1\u001b[0m\n\u001b[1;33m    print(\"Невідповідність відкритих та закритих лапок!)\u001b[0m\n\u001b[1;37m                                                        ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m EOL while scanning string literal\n"
     ]
    }
   ],
   "source": [
    "print(\"Невідповідність відкритих та закритих лапок!)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Семантичні** - це помилки в логіці роботи програми, які можна виявити тільки за результатами роботи скрипта. Як правило, інтерпретатор не попереджає про наявність помилки. А програма буде виконуватися, оскільки не містить синтаксичних помилок. Такі помилки досить важко виявити і виправити.\n",
    "\n",
    "**Помилки часу виконання** - це помилки, які виникають під час роботи скрипта. Причиною є події, які не передбачені програмістом. Класичним прикладом служить ділення на нуль:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "ZeroDivisionError",
     "evalue": "division by zero",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-3-29fd4ab02719>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mtest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m/\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m4\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-3-29fd4ab02719>\u001b[0m in \u001b[0;36mtest\u001b[1;34m(x, y)\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mtest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m/\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m4\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mZeroDivisionError\u001b[0m: division by zero"
     ]
    }
   ],
   "source": [
    "def test(x, y):\n",
    "    return x/y\n",
    "print(test(4, 0))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "В мові Python винятки порушуються не тільки при помилці, але і як повідомлення про настання будь-яких подій. Наприклад, метод *index()* збуджує виняток *ValueError*, якщо шуканий фрагмент не входить в рядок:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"String\".index(\"S\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "substring not found",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-8-b68a44906112>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;34m\"String\"\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"s\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m: substring not found"
     ]
    }
   ],
   "source": [
    "\"String\".index(\"s\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.2. Оброблення винятків\n",
    "\n",
    "Для оброблення винятків призначена інструкція **try**. Формат інструкції:\n",
    "\n",
    "**try:**\n",
    "\n",
    "$\\phantom{w}$ <БЛОК, В ЯКОМУ ПЕРЕХОПЛЮЮТЬСЯ ВИНЯТКИ>\n",
    "\n",
    "**except** <ВИНЯТОК_1> **as** <ОБ'ЄКТ ВИНЯТКУ>**:**\n",
    "\n",
    "$\\phantom{w}$ <БЛОК, ЩО ВИКОНУЄТЬСЯ ПРИ ЗБУДЖЕННІ ВИНЯТКУ>\n",
    "\n",
    "**. . .**\n",
    "\n",
    "**except** <ВИНЯТОК_N> **as** <ОБ'ЄКТ ВИНЯТКУ>**:**\n",
    "\n",
    "$\\phantom{w}$ <БЛОК, ЩО ВИКОНУЄТЬСЯ ПРИ ЗБУДЖЕННІ ВИНЯТКУ>\n",
    "\n",
    "**else:** <БЛОК, ЩО ВИКОНУЄТЬСЯ, ЯКЩО ВИНЯТКУ НЕ ВИНИКЛО>\n",
    "\n",
    "**finally:** <БЛОК, ЩО ВИКОНУЄТЬСЯ В БУДЬ-ЯКОМУ ВИПАДКУ>\n",
    "\n",
    "Інструкції, в яких перехоплюються винятки, повинні бути розташовані всередині блоку **try**. У блоці **except** в параметрі <Виняток_1> вказується клас оброблюваного винятку.\n",
    "\n",
    "Наприклад, щоб обробити виняток, що виникає при діленні на нуль:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Обробили ділення на 0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "try: # Перехоплюється виняток\n",
    "    х = 1/0 # Помилка: ділення на 0\n",
    "except ZeroDivisionError: # Вказуємо клас винятку\n",
    "    print (\"Обробили ділення на 0\")\n",
    "х=0\n",
    "print(х)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Якщо в блоці **try** згенеровано виняток, то управління передається блоку **except**. У разі, якщо виключення не відповідає зазначеному класу, управління передається наступному блоку **except**. Якщо жоден блок **except** не відповідає винятку, то виняток \"спливає\" до обробника більш високого рівня. Якщо виняток ніде не обробляється в програмі, то управління передається обробнику за замовчуванням, який зупиняє виконання програми і виводить стандартну інформацію про помилку. Таким чином, в обробнику може бути кілька блоків **except** з різними класами винятків. Крім того, один обробник можна вкласти в інший.\n",
    "\n",
    "Детальніше про винятки можна прочитати в книжці."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Запитання для самоконтролю\n",
    "\n",
    "1. Поняття помилки.\n",
    "2. Що таке виняткові ситуації і яким чином здійснюється їх оброблення у Python?\n",
    "3. Блоки try - except.\n",
    "4. Атрибути винятків, ініціювання винятків.\n",
    "5. Для чого використовується гілка finally в інструкції try?\n",
    "6. Чи можна гілку finally поєднувати з гілками except?\n",
    "7. Які два класи виняткових ситуацій наявні в Python?\n",
    "8. Який із класів виняткових ситуацій рекомендується використовувати у програмах?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "colab": {
   "collapsed_sections": [],
   "name": "SMS_model_2021-to-Colab.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
